File: dsa/hash_table/454_4Sum_II.java
package hash_table;
import java.util.*;

class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> mapVal = new HashMap<Integer, Integer>();
        int result = 0;

        for (int i = 0; i < nums1.length; i++){
            for (int j = 0; j < nums2.length; j++){
                int tempSum = nums1[i] + nums2[j];
                if (mapVal.containsKey(tempSum) == true){
                    mapVal.replace(tempSum, mapVal.get(tempSum) + 1);
                }
                else{
                    mapVal.put(tempSum, 1);
                }
            }
        }

        for (int k = 0; k < nums3.length; k++){
            for (int l = 0; l < nums4.length; l++){
                int tempSum = -(nums3[k] + nums4[l]);
                if (mapVal.containsKey(tempSum) == true){
                    result += mapVal.get(tempSum);
                }
            }
        }
        return result;
    }
}

File: dsa/6.hash_table/454_4Sum_II.py
import typing as t

class Solution:
    def fourSumCount(self, nums1: t.List[int], nums2: t.List[int], nums3: t.List[int], nums4: t.List[int]) -> int:
        freq_sum_dict: t.Dict[int] = {}
        result: int = 0

        for ind1 in range(0, len(nums1), 1):
            for ind2 in range(0, len(nums2), 1):
                temp_sum: int = -(nums1[ind1] + nums2[ind2])
                if temp_sum not in freq_sum_dict.keys():
                    freq_sum_dict[temp_sum] = 1
                else:
                    freq_sum_dict[temp_sum] = freq_sum_dict.get(temp_sum) + 1
        
        for ind3 in range(0, len(nums3), 1):
            for ind4 in range(0, len(nums4), 1):
                temp_sum: int = nums3[ind3] + nums4[ind4]
                if temp_sum in freq_sum_dict.keys():
                    result += freq_sum_dict.get(temp_sum)
        
        return result
